src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/CertificatePal.cs

```diff
@@ -480,20 +469,57 @@ private ICertificatePal CopyWithPrivateKey(SecKeyPair keyPair)
 
             SafeKeychainHandle keychain = Interop.AppleCrypto.SecKeychainItemCopyKeychain(keyPair.PrivateKey);
 
+            // If we're using a key already in a keychain don't add the certificate to that keychain here,
+            // do it in the temporary add/remove in the shim.
+            SafeKeychainHandle cloneKeychain = SafeTemporaryKeychainHandle.InvalidHandle;
+
             if (keychain.IsInvalid)
             {
                 keychain = Interop.AppleCrypto.CreateTemporaryKeychain();
+                cloneKeychain = keychain;
+            }
+
+            // Because SecIdentityRef only has private constructors we need to have the cert and the key
+            // in the same keychain.  That almost certainly means we're going to need to add this cert to a
+            // keychain, and when a cert that isn't part of a keychain gets added to a keychain then the
+            // interior pointer of "what keychain did I come from?" used by SecKeychainItemCopyKeychain gets
+            // set. That makes this function have side effects, which is not desired.
+            //
+            // It also makes reference tracking on temporary keychains broken, since the cert can
+            // DangerousRelease a handle it didn't DangerousAddRef on.  And so CopyWithPrivateKey makes
+            // a temporary keychain, then deletes it before anyone has a chance to (e.g.) export the
+            // new identity as a PKCS#12 blob.
+            //
+            // Solution: Clone the cert, like we do in Windows.
+            SafeSecCertificateHandle tempHandle;
+
+            {
+                byte[] export = RawData;
+                const bool exportable = false;
+                SafeSecIdentityHandle identityHandle;
+                tempHandle = Interop.AppleCrypto.X509ImportCertificate(
+                    export,
+                    X509ContentType.Cert,
+                    SafePasswordHandle.InvalidHandle,
+                    cloneKeychain,
```

Comment:
I'm confused as to the purpose (or maybe just the naming?) of `cloneKeychain`.  It looks like it's either `InvalidHandle` or `keychain` based on whether the keychain was originally invalid?  So it's never actually a clone, it's either an invalid handle or the exact same reference.  Is the variable even necessary?  Could there just be a bool that's set and then this is `wasInvalid ? keychain : InvalidHandle`?